internal/testutils/testserver.go

```diff
@@ -51,55 +51,87 @@ func (t TestServer) FaultyClient() *http.Client {
-	ValidateRequestFunc func(*http.Request) // Function to validate incoming requests.
-}
-
 // NewHTTPTestServer creates a new HTTP test server with the specified responses for each HTTP method.
-func NewHTTPTestServer(t *testing.T, responses []ServerResponses) *TestServer {
+func NewHTTPTestServer(t *testing.T, responses []ResponseDef) *TestServer {
 	testServer := &TestServer{}
 	handler := func(rw http.ResponseWriter, req *http.Request) {
 		testServer.calls++
-		// Special case: if only one response is specified, this will be reused
-		// for each and every call
-		if len(responses) == 1 {
-			if res, found := responses[0][req.Method]; found {
-				if res.ValidateRequestFunc != nil {
-					res.ValidateRequestFunc(req)
-				}
-				rw.WriteHeader(res.ResponseCode)
-				_, _ = rw.Write([]byte(res.ResponseBody)) // nosemgrep: go.lang.security.audit.xss.no-direct-write-to-responsewriter.no-direct-write-to-responsewriter
-			} else {
-				t.Errorf("unexpected HTTP method call: %s", req.Method)
-			}
-			return
+		if len(responses) <= testServer.calls-1 {
+			t.Errorf("Exceeded number of calls to test server (expected: %d)", len(responses))
 		}
 
-		if len(responses) <= testServer.calls-1 {
-			t.Fail()
-			t.Fatalf("Exceeded number of calls to test server (expected: %d)", len(responses))
+		responseDef := responses[testServer.calls-1]
+		methodFuncs := map[string]func(*testing.T, *http.Request) Response{
+			http.MethodGet:    responseDef.GET,
+			http.MethodPost:   responseDef.POST,
+			http.MethodPut:    responseDef.PUT,
+			http.MethodDelete: responseDef.DELETE,
 		}
 
-		if res, found := responses[testServer.calls-1][req.Method]; found {
-			if res.ValidateRequestFunc != nil {
-				res.ValidateRequestFunc(req)
-			}
-			rw.WriteHeader(res.ResponseCode)
-			_, _ = rw.Write([]byte(res.ResponseBody)) // nosemgrep: go.lang.security.audit.xss.no-direct-write-to-responsewriter.no-direct-write-to-responsewriter
-		} else {
-			t.Errorf("unexpected HTTP method call: %s", req.Method)
+		handlerFunc, found := methodFuncs[req.Method]
+		if !found {
+			rw.WriteHeader(http.StatusMethodNotAllowed)
+			_, _ = rw.Write([]byte("Method not supported")) // nosemgrep: go.lang.security.audit.xss.no-direct-write-to-responsewriter.no-direct-write-to-responsewriter
+			return
 		}
+		if handlerFunc == nil {
+			t.Errorf("No %s method defined for server call nr. %d", req.Method, testServer.calls)
+			return
+		}
+
+		response := handlerFunc(t, req)
+		rw.WriteHeader(response.ResponseCode)
+		_, _ = rw.Write([]byte(response.ResponseBody)) // nosemgrep: go.lang.security.audit.xss.no-direct-write-to-responsewriter.no-direct-write-to-responsewriter
+		responseDef.Validate(t, req)
 	}
 	testServer.Server = httptest.NewServer(http.HandlerFunc(handler))
 	return testServer
 }
 
+type Response struct {
+	ResponseCode int
+	ResponseBody string
+}
+
+type ResponseDef struct {
+	GET             func(*testing.T, *http.Request) Response
```

Comment:
nit: It might be a bit confusing that you can define several methods for a single response - and the server will answer to whatever was requested. Can't think of a better way than going back to a map[string]func though...

Reply:
I thought of including a check that panics if you define more than one, wdyt?

Reply:
That'd feel in line with what we said above about failing if unexpected calls happen. I don't know if there'd ever be a case, if I want to define an either/or kind of reply for a test... 🤔 Probably not. 
Maybe t.Fatal would be cleaner than just panicing?

Reply:
not sure, actually panic makes the most sense to me.
actually also in the case where a handler is `!found` it should panic.
It's no test error, it's just wrongly used and could never work.

That's also how https://github.com/matryer/moq are handling these situations. if something not defined on your mock is called, it panics.

are you fine with that?